gitium/functions.php

```diff
@@ -206,10 +206,45 @@ function gitium_commit_and_push_gitignore_file( $path = '' ) {
 	gitium_merge_and_push( $commit );
 }
 
+if ( ! function_exists( 'gitium_acquire_merge_lock' ) ) :
+	function gitium_acquire_merge_lock() {
+		$gitium_lock_path   = apply_filters( 'gitium_lock_path', ABSPATH . 'wp-content/.gitium-lock' );
+		$gitium_lock_handle = fopen( $gitium_lock_path, 'w+' );
+
+		if ( flock( $gitium_lock_handle, LOCK_EX | LOCK_NB ) ) {
+			return array( $gitium_lock_path, $gitium_lock_handle );
+		} else {
+			$lock_timeout    = intval( ini_get( 'max_execution_time' ) ) > 0 ? intval( ini_get( 'max_execution_time' ) ) / 2 : 15;
+			$lock_timeout_ms = 10;
+			$lock_retries    = 0;
+			while ( $lock_retries * $lock_timeout_ms < $lock_timeout * 1000 ) {
+				if ( flock( $gitium_lock_handle, LOCK_EX | LOCK_NB ) ) {
+					return array( $gitium_lock_path, $gitium_lock_handle );
+				}
+				usleep( $lock_timeout_ms * 1000 );
+				$lock_retries++;
+			}
+			return false; // timeout
+		}
+	}
+endif;
+
+if ( ! function_exists( 'gitium_release_merge_lock' ) ) :
+	function gitium_release_merge_lock( $lock ) {
+		list( $gitium_lock_path, $gitium_lock_handle ) = $lock;
+		flock( $gitium_lock_handle, LOCK_UN );
+		fclose( $gitium_lock_handle );
+		unlink( $gitium_lock_path );
+	}
+endif;
+
 // Merges the commits with remote and pushes them back
 function gitium_merge_and_push( $commits ) {
 	global $git;
 
+	$lock = gitium_acquire_merge_lock()
+		or trigger_error( 'Error when gitium lock was acquired because of a timeout', E_USER_ERROR );
```

Comment:
This is not an error (eg. request should not be halted) it should rather be a warning condition.